File: Code tham khảo/SIL/SilAdapterBasicPython/SUT/Application_Python/logic.py
from iolayer import State, get_sensor_value1, get_sensor_value2, get_sensor_value3, set_heating_state

HEATING_THRESHOLD = 18.0
COOLING_THRESHOLD = 23.0
BOOL_TO_ON_OFF_MAP = {True: State.On,
                      False: State.Off}

write_to_console_counter = 0

def get_heating_state(temperature: float) -> State:
    return BOOL_TO_ON_OFF_MAP[temperature < HEATING_THRESHOLD]
    

def get_cooling_state(temperature: float) -> State:
    return BOOL_TO_ON_OFF_MAP[temperature > COOLING_THRESHOLD]


def print_status_to_console(sensor1: float, sensor2: float, sensor3: float, heating_state: State, cooling_state: State) -> None:
    print("Sensor1: " + str(sensor1) +
          ", Sensor2: " + str(sensor2) +
          ", Sensor3: " + str(sensor3) +
          ", heating is " + heating_state.name +
          ", cooling is " + cooling_state.name)


def calc_temperature(value1: float, value2: float, value3: float) -> float:
    return (value1 + value2 + value3) / 3


def control_room_temperature():
    global write_to_console_counter

    sensor1 = get_sensor_value1()
    sensor2 = get_sensor_value2()
    sensor3 = get_sensor_value3()

    temperature = calc_temperature(sensor1, sensor2, sensor3) 

    heating_state = get_heating_state(temperature)
    cooling_state = get_cooling_state(temperature)
    
    set_heating_state(heating_state, cooling_state)

    if write_to_console_counter >= 9:
        print_status_to_console(sensor1, sensor2, sensor3, heating_state, cooling_state)
        write_to_console_counter = 0
    else:
        write_to_console_counter += 1

File: Code tham khảo/SIL/SilAdapterBasicCpp/SUT/Application_C/src/Logic.c
/*
 * This file contains the actual application logic.
 */

#include <stdio.h>

#include "IO.h"
#include "Logic.h"

#define HEATING_THRESHOLD 18.0
#define COOLING_THRESHOLD 23.0

enum State GetHeatingState(double temperature) { return (temperature < HEATING_THRESHOLD) ? STATE_ON : STATE_OFF; }

enum State GetCoolingState(double temperature) { return (temperature > COOLING_THRESHOLD) ? STATE_ON : STATE_OFF; }

void PrintStatusToConsole(double sensor1, double sensor2, double sensor3, enum State heatingState, enum State coolingState)
{
  printf("Sensor1: %.2f, Sensor2: %.2f, Sensor3: %.2f, heating is %s, cooling is %s\r\n",
      sensor1,
      sensor2,
      sensor3,
      (heatingState == STATE_ON) ? "ON" : "OFF",
      (coolingState == STATE_ON) ? "ON" : "OFF");
}

double CalcTemperature(double value1, double value2, double value3) { return (value1 + value2 + value3) / 3.0; }

void ControlRoomTemperature()
{
  static int writeToConsoleCounter = 0;

  const double sensor1 = GetSensorValue1();
  const double sensor2 = GetSensorValue2();
  const double sensor3 = GetSensorValue3();

  const double temperature = CalcTemperature(sensor1, sensor2, sensor3);

  const enum State heatingState = GetHeatingState(temperature);
  const enum State coolingState = GetCoolingState(temperature);

  SetHeatingState(heatingState, coolingState);

  if (writeToConsoleCounter++ >= 9)
  {
    PrintStatusToConsole(sensor1, sensor2, sensor3, heatingState, coolingState);
    writeToConsoleCounter = 0;
  }
}
